src/en/mangahen/src/eu/kanade/tachiyomi/extension/en/mangahen/MangaHen.kt

```diff
@@ -0,0 +1,170 @@
+package eu.kanade.tachiyomi.extension.en.mangahen
+
+import eu.kanade.tachiyomi.network.GET
+import eu.kanade.tachiyomi.source.model.FilterList
+import eu.kanade.tachiyomi.source.model.MangasPage
+import eu.kanade.tachiyomi.source.model.Page
+import eu.kanade.tachiyomi.source.model.SChapter
+import eu.kanade.tachiyomi.source.model.SManga
+import eu.kanade.tachiyomi.source.online.HttpSource
+import eu.kanade.tachiyomi.util.asJsoup
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.Request
+import okhttp3.Response
+import org.jsoup.nodes.Element
+import rx.Observable
+
+class MangaHen : HttpSource() {
+
+    override val name = "MangaHen"
+
+    override val baseUrl = "https://manga-hen.com"
+
+    private val advSearchURL = "$baseUrl/advanced-search"
+
+    override val lang = "en"
+
+    override val supportsLatest = true
+
+    override val client = network.cloudflareClient
+
+    // Popular
+    override fun popularMangaRequest(page: Int): Request {
+        return GET("$advSearchURL/?search=1&type=0&sort=1&page=$page", headers)
+    }
+
+    override fun popularMangaParse(response: Response): MangasPage {
+        val doc = response.asJsoup()
+
+        val mangas = doc.select("a[href^=/manga/]").map(::popularMangaFromElement)
+
+        val hasNextPage = doc.select("a[href*=page]").any { it.text().isBlank() }
+
+        return MangasPage(mangas, hasNextPage)
+    }
+
+    private fun popularMangaFromElement(element: Element): SManga {
+        return SManga.create().apply {
+            title = element.selectFirst("h2")!!.ownText()
+            setUrlWithoutDomain(element.absUrl("href"))
+            thumbnail_url = element.selectFirst("img")!!.absUrl("src")
+        }
+    }
+
+    // Latest
+    override fun latestUpdatesRequest(page: Int): Request {
+        return GET("$advSearchURL/?search=1&type=0&sort=2&page=$page", headers)
+    }
+
+    override fun latestUpdatesParse(response: Response): MangasPage = popularMangaParse(response)
+
+    // Search
+
+    private fun tagSearch(tag: String, tagsList: List<String>): String? {
+        val index = (tagsList.indexOf(tag) + 1).toString()
+        return if (index != "-1") index else null
+    }
+
+    private fun tagsList(): List<String> {
+        val request = GET(advSearchURL, headers)
+
+        val response = client.newCall(request).execute()
+        val tagsList = response.asJsoup().select("li[onclick=updateTag(this)]").map { it.ownText().lowercase() }
+
+        return tagsList
+    }
+
+    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
+        val includeTags = mutableListOf<String>()
+        val excludeTags = mutableListOf<String>()
+
+        val tagsList = tagsList()

```

Comment:
getting tags everytime isn't ideal, cache them in a variable.

also you can get them and populate FilterList dynamically, so users will not need to write them in text filter. [example implementation where genres are dynamically fetched ](https://github.com/keiyoushi/extensions-source/blob/main/src/all/galaxy/src/eu/kanade/tachiyomi/extension/all/galaxy/Galaxy.kt#L94-L157)

Reply:
Thank you for the suggestion! I've changed the code according to your suggestion, is it correct?